syntax = "proto3";

// Định nghĩa các message và service cho Tickets

message Ticket {
  int32 ticket_id = 1;          // ID của vé
  ỉnt32 event_id = 2;          // ID sự kiện
  string ticket_type = 3;       // Loại vé
  float ticket_price = 4;       // Giá vé
  int32 total_quantity = 5;     // Tổng số lượng vé
  int32 available_quantity = 6; // Số lượng vé còn lại
}

// Request để thêm hoặc cập nhật vé
message TicketRequest {
  int32 event_id = 1;          // ID sự kiện
  string ticket_type = 2;       // Loại vé
  float ticket_price = 3;       // Giá vé
  int32 total_quantity = 4;     // Tổng số lượng vé
  int32 available_quantity = 5; // Số lượng vé còn lại
}

// Response cho một vé cụ thể
message TicketResponse {
  Ticket ticket = 1;            // Đối tượng vé
}

// Request để lấy tất cả vé
message GetAllTicketsRequest {}

// Response chứa danh sách tất cả vé
message GetAllTicketsResponse {
  repeated Ticket tickets = 1;  // Danh sách vé
}

// Request để xóa một vé theo ID
message DeleteTicketRequest {
  int32 ticket_id = 1;          // ID của vé cần xóa
}

// Response sau khi xóa vé
message DeleteTicketResponse {
  bool success = 1;             // Trạng thái thành công hay thất bại
}



// Service để quản lý vé
service TicketService {
  // RPC để thêm một vé
  rpc AddTicket (TicketRequest) returns (TicketResponse);

  // RPC để cập nhật một vé
  rpc UpdateTicket (Ticket) returns (TicketResponse);

  // RPC để lấy tất cả vé
  rpc GetAllTickets (GetAllTicketsRequest) returns (GetAllTicketsResponse);

  // RPC để xóa một vé
  rpc DeleteTicket (DeleteTicketRequest) returns (DeleteTicketResponse);
}


// Định nghĩa các message và service cho Orders

message Order {
  int32 order_id = 1;            // ID của đơn hàng
  string customer_name = 2;      // Tên khách hàng
  repeated OrderItem items = 3;  // Danh sách các vé trong đơn hàng
  float total_price = 4;         // Tổng giá trị đơn hàng
  string status = 5;             // Trạng thái của đơn hàng (e.g., pending, confirmed, cancelled)
}

message OrderItem {
  int32 ticket_id = 1;           // ID của vé
  int32 quantity = 2;            // Số lượng vé
  float price = 3;               // Giá của vé
}

// Request để thêm hoặc cập nhật đơn hàng
message OrderRequest {
  string customer_name = 1;      // Tên khách hàng
  repeated OrderItem items = 2;  // Danh sách các vé trong đơn hàng
  float total_price = 3;
}

// Response cho một đơn hàng cụ thể
message OrderResponse {
  Order order = 1;               // Đối tượng đơn hàng
}

// Request để lấy tất cả đơn hàng
message GetAllOrdersRequest {}

// Response chứa danh sách tất cả đơn hàng
message GetAllOrdersResponse {
  repeated Order orders = 1;     // Danh sách đơn hàng
}

// Request để xóa một đơn hàng theo ID
message DeleteOrderRequest {
  int32 order_id = 1;            // ID của đơn hàng cần xóa
}

// Response sau khi xóa đơn hàng
message DeleteOrderResponse {
  bool success = 1;              // Trạng thái thành công hay thất bại
}

// Service để quản lý đơn hàng
service OrderService {
  // RPC để thêm một đơn hàng
  rpc AddOrder (OrderRequest) returns (OrderResponse);

  // RPC để cập nhật một đơn hàng
  rpc UpdateOrder (Order) returns (OrderResponse);

  // RPC để lấy tất cả đơn hàng
  rpc GetAllOrders (GetAllOrdersRequest) returns (GetAllOrdersResponse);

  // RPC để xóa một đơn hàng
  rpc DeleteOrder (DeleteOrderRequest) returns (DeleteOrderResponse);
}